### PR TITLE
Improve network handling

### PR DESCRIPTION
**Pull request by Temikus**
I was frustrated for a while that network didn't accept models and finally fixed this.
Did quite a bit of drive-by fixing while I was at it. This is also somewhat related to #183 

- Made network-config handling model-oriented
- Detangled network handling from the main server.insert routine
- Abstracted network config generation to the Model
- Had to make the network creation synchronous because there's no status in the API, this now means we can depend on the network to exist without trying and rolling back an expensive operation.

+ Small fixes and improvements

P.S. Integration tests added. #happening \o/

### CODE REVIEW

**Comment by Temikus** [comment]
Partially related to #183, should be a bit easier to fix now.

File: test/integration/compute/test_compute_networks_collection.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    assert_equal(my_network_name, my_network.name, "My network should have the provided name")
+
+    # It should also be visible when listing addresses
+    assert_includes(client.networks.all.map(&:name), my_network_name)
+
+    # Be aware that although the address resource is created, it might not yet
+    # have an ip address. You can poll until the address has been assigned.
+    my_network.wait_for { !my_network.ipv4_range.nil? }
+    assert_match(/\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\/\d{1,2}/,
```
Comment (Temikus) [inline test/integration/compute/test_compute_networks_collection.rb]: Irrelevant for minitest.

File: lib/fog/compute/google/requests/insert_server.rb [hunk 2]
```diff
@@ -110,6 +110,33 @@ def handle_disks(options)
...
+          access_config = { "type" => "ONE_TO_ONE_NAT", "name" => "External NAT" }
+          # If natIP is undefined VM will get an IP from a shared ephemeral IP address pool
+          if options.key? "externalIp"
+            access_config["natIP"] = options.delete "externalIp"
+            # If :external_ip is set to 'false', do not allow _any_ external networking
+            access_config = nil if access_config["natIP"] == false
+          end
+
+          # If no networking options are specified, assume default network
+          if options.key? "network"
```
Comment (Temikus) [inline lib/fog/compute/google/requests/insert_server.rb]: Imo this is less readable.

File: tasks/console.rake [hunk 3]
```diff
@@ -5,6 +5,13 @@ desc "Project IRB console"
 task :console do
   require "bundler"
   Bundler.require(:default, :development)
+
+  # Reload helper to avoid resetting the environment when debugging
+  def reload!
+    files = $LOADED_FEATURES.select { |feat| feat =~ /\/fog-google\// }
```
Comment (Temikus) [inline tasks/console.rake]: Fixed the style.

File: examples/precreated_client.rb [hunk 4]
```diff
@@ -1,3 +1,14 @@
+# All examples presume that you have a ~/.fog credentials file set up.
```
Comment (Temikus) [inline examples/precreated_client.rb]: Fixed the config, this is not something that should be addressed by the linter.

File: examples/network.rb [hunk 5]
```diff
@@ -1,11 +1,25 @@
+# All examples presume that you have a ~/.fog credentials file set up.
```
Comment (Temikus) [inline examples/network.rb]: Fixed the config, this is not something that should be addressed by the linter.

File: test/integration/compute/test_compute_networks_collection.rb (hunk 1 above)
Comment (icco) [inline test/integration/compute/test_compute_networks_collection.rb]: I think the idea here is %r just makes it easier to read?

Comment (icco) [inline test/integration/compute/test_compute_networks_collection.rb]: Ah, nevermind, saw your comment below. I don't care :D

File: lib/fog/compute/google/models/network.rb [hunk 6]
```diff
@@ -21,7 +21,9 @@ def save
 
           data = service.insert_network(identity, ipv4_range, attributes)
           operation = Fog::Compute::Google::Operations.new(:service => service).get(data.body["name"])
-          operation.wait_for { !pending? }
+          # Since network has no "state" we can query, we have to wait for the operation to finish
+          # TODO: change back to async when
```
Comment (icco) [inline lib/fog/compute/google/models/network.rb]: This TODO doesn't make sense?

File: lib/fog/compute/google/requests/insert_server.rb (hunk 2 above)
Comment (icco) [inline lib/fog/compute/google/requests/insert_server.rb]: +1 I thought we had turned that off.

File: test/integration/compute/test_compute_networks_collection.rb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      :disks => [
+        :boot => true,
+        :autoDelete => true,
+        :initializeParams => {
+          :sourceImage => "projects/debian-cloud/global/images/family/debian-8"
+        }
+      ],
+      :network => my_network
+    )
+    my_server.wait_for { my_server.state != "PROVISIONING" }
```
Comment (icco) [inline test/integration/compute/test_compute_networks_collection.rb]: This should be `!pending?` I think

File: test/integration/compute/test_compute_networks_collection.rb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    assert_equal(
+      "ONE_TO_ONE_NAT",
+      my_server.network_interfaces[0]["accessConfigs"][0]["type"],
+      "Address should now be in use"
+    )
+
+    assert_equal(
+      "External NAT",
+      my_server.network_interfaces[0]["accessConfigs"][0]["name"],
+      "Address should now be in use"
```
Comment (icco) [inline test/integration/compute/test_compute_networks_collection.rb]: These failure messages don't make much sense to me. They should probably describe what is wrong.

File: test/integration/compute/test_compute_networks_collection.rb (hunk 7 above)
Comment (Temikus) [inline test/integration/compute/test_compute_networks_collection.rb]: There's no PENDING state for a server but this gives me an idea...

File: test/integration/compute/test_compute_networks_collection.rb (hunk 8 above)
Comment (Temikus) [inline test/integration/compute/test_compute_networks_collection.rb]: You're 100% right, missed that.

File: lib/fog/compute/google/models/network.rb (hunk 6 above)
Comment (Temikus) [inline lib/fog/compute/google/models/network.rb]: Yep, thanks. Will fix.

**Comment by Temikus** [comment]
@icco Thanks for the comments! Addressed, PTAL.

**Pull request merged**